File: go/dsa/list/CircularLinkedList/CircularLinkedList.go
package circularlinkedlist

import "fmt"

// ListNode representa um nó em uma lista ligada circular.
type ListNode struct {
	Next *ListNode // Ponteiro para o próximo nó na lista
	Data int       // Dados armazenados no nó
}

// CircularLinkedList representa uma lista ligada circular.
type CircularLinkedList struct {
	last   *ListNode // Ponteiro para o último nó da lista (que aponta para o primeiro nó)
	length int       // Número de nós na lista
}

// NewCircularLinkedList cria uma nova lista ligada circular vazia.
func NewCircularLinkedList() *CircularLinkedList {
	return &CircularLinkedList{}
}

// Length retorna o número de nós na lista.
func (cl *CircularLinkedList) Length() int {
	return cl.length
}

// IsEmpty verifica se a lista está vazia.
func (cl *CircularLinkedList) IsEmpty() bool {
	return cl.length == 0
}

// CreateCircularLinkedList cria uma lista ligada circular com quatro nós.
func (cl *CircularLinkedList) CreateCircularLinkedList() {
	// Cria quatro nós
	first := &ListNode{Data: 1}
	second := &ListNode{Data: 5}
	third := &ListNode{Data: 10}
	fourth := &ListNode{Data: 15}

	// Conecta os nós entre si para formar um círculo
	first.Next = second
	second.Next = third
	third.Next = fourth
	fourth.Next = first

	// Define o último nó como o quarto nó e ajusta o comprimento da lista
	cl.last = fourth
	cl.length = 4
}

// Display exibe todos os nós da lista ligada circular.
func (cl *CircularLinkedList) Display() {
	// Se a lista estiver vazia, não faz nada
	if cl.last == nil {
		return
	}

	// Começa pelo primeiro nó (o nó após o último)
	first := cl.last.Next
	// Itera sobre os nós e imprime seus valores
	for first != cl.last {
		fmt.Printf("%d --> ", first.Data)
		first = first.Next
	}
	// Imprime o valor do último nó para completar o ciclo
	fmt.Println(first.Data)
}

// InsertFirst insere um novo nó no início da lista ligada circular.
func (cl *CircularLinkedList) InsertFirst(Data int) {
	// Cria um novo nó com os dados fornecidos
	temp := &ListNode{Data: Data}
	if cl.last == nil {
		// Se a lista estiver vazia, o novo nó aponta para si mesmo e se torna o último nó
		cl.last = temp
		cl.last.Next = cl.last
	} else {
		// Caso contrário, insere o novo nó após o último nó e o conecta ao primeiro
		temp.Next = cl.last.Next
		cl.last.Next = temp
	}
	// Incrementa o comprimento da lista
	cl.length++
}

// InsertLast insere um novo nó no final da lista ligada circular.
func (cl *CircularLinkedList) InsertLast(Data int) {
	// Cria um novo nó com os dados fornecidos
	temp := &ListNode{Data: Data}
	if cl.last == nil {
		// Se a lista estiver vazia, o novo nó aponta para si mesmo e se torna o último nó
		cl.last = temp
		cl.last.Next = cl.last
	} else {
		// Caso contrário, insere o novo nó após o último nó e o atualiza como o novo último nó
		temp.Next = cl.last.Next
		cl.last.Next = temp
		cl.last = temp
	}
	// Incrementa o comprimento da lista
	cl.length++
}

// RemoveFirst remove o primeiro nó da lista ligada circular e retorna o nó removido.
func (cl *CircularLinkedList) RemoveFirst() *ListNode {
	// Verifica se a lista está vazia
	if cl.IsEmpty() {
		panic("Circular Singly Linked List is already empty")
	}

	// Temporariamente armazena o nó que será removido (o primeiro nó)
	temp := cl.last.Next
	if cl.last.Next == cl.last {
		// Se houver apenas um nó, a lista se torna vazia
		cl.last = nil
	} else {
		// Caso contrário, o segundo nó se torna o primeiro nó
		cl.last.Next = temp.Next
	}
	// Desconecta o nó removido e diminui o comprimento da lista
	temp.Next = nil
	cl.length--
	return temp // Retorna o nó removido
}

File: csharp/src/dsa/list/CircularLinkedList.cs
namespace dsa.list;

public class CircularLinkedList
{
    // static void Main()
    // {
    //     CircularLinkedList cll = new();
    //     cll.createCircularLinkedList();
    //     cll.display();
    // }

    // A classe CircularLinkedList tem dois campos: 'last' que aponta para o último nó da lista e '_length' que armazena o tamanho da lista
    public ListNode last;
    public int _length;

    // A classe ListNode representa um nó na lista. Cada nó tem um campo 'data' que armazena o valor do nó e um campo 'next' que aponta para o próximo nó na lista
    public class ListNode
    {
        public ListNode next;
        public int data;

        // Construtor da classe ListNode
        public ListNode(int data) => this.data = data;
    }

    // Construtor da classe CircularLinkedList
    public CircularLinkedList()
    {
        last = null;
        this._length = 0;
    }

    // Método para obter o tamanho da lista
    public int length() => _length;

    // Método para verificar se a lista está vazia
    public bool isEmpty() => _length == 0;

    // Método para criar uma lista circularmente encadeada com quatro nós
    public void createCircularLinkedList()
    {
        ListNode first = new ListNode(1);
        ListNode second = new ListNode(5);
        ListNode third = new ListNode(10);
        ListNode fourth = new ListNode(15);

        first.next = second;
        second.next = third;
        third.next = fourth;
        fourth.next = first;

        last = fourth;
    }

    // Método para exibir a lista
    public void display()
    {
        if (last == null)
            return;

        ListNode first = last.next;
        while (first != last)
        {
            Console.Write(first.data + " --> ");
            first = first.next;
        }
        Console.WriteLine(first.data);
    }

    // Método para inserir um nó no início da lista
    public void insertFirst(int data)
    {
        ListNode temp = new ListNode(data);
        if (last == null)
            last = temp;

        else
            temp.next = last.next;

        last.next = temp;
        _length++;
    }

    // Método para inserir um nó no final da lista
    public void insertLast(int data)
    {
        ListNode temp = new ListNode(data);
        if (last == null)
        {
            last = temp;
            last.next = last;
        }
        else
        {
            temp.next = last.next;
            last.next = temp;
            last = temp;
        }
        _length++;
    }

    // Método para remover o primeiro nó da lista
    public ListNode removeFirst()
    {
        if (isEmpty())
            throw new NullReferenceException("Circular Singly Linked List is already empty");

        ListNode temp = last.next;
        if (last.next == last)
            last = null;

        else
            last.next = temp.next;

        temp.next = null;
        _length--;
        return temp;
    }
}
